[flang] Added fir.is_contiguous_box and fir.box_total_elements ops.

vzakhari opened the pull request:
These are helper operations to aid with expanding of fir.pack_array.

vzakhari committed:
[flang] Added fir.is_contiguous_box and fir.box_total_elements ops.

These are helper operations to aid with expanding of fir.pack_array.

vzakhari committed:
[flang] Expand new FIR box operations into runtime calls.

jeanPerier commented on flang/lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -4671,6 +4671,79 @@ void fir::UnpackArrayOp::getEffects(
+    rewriter.replaceOpWithNewOp<mlir::arith::ConstantOp>(
+        op, op.getType(), rewriter.getIntegerAttr(op.getType(), 1));
+    return mlir::success();
+  }
```
Result of fir.embox without slices can also fold to true (there are other cases, but it is a good start and some fir::isSimplyConintiguous(box) helper could be added later to try to walk back more, which could be helpful after inlining).

jeanPerier commented on flang/lib/Optimizer/Transforms/SimplifyFIROperations.cpp:
```diff
@@ -0,0 +1,157 @@
+  fir::populateSimplifyFIROperationsPatterns(patterns,
+                                             preferInlineImplementation);
+  if (mlir::failed(
+          mlir::applyPartialConversion(module, target, std::move(patterns)))) {
```
As I understand it, dialect conversions are more expensive than applyPatternsAndFoldGreedily that could be used here given there is no operation result type change.

The illegal op check is nice, but I am not sure it is worth using it (codegen will complain anyway).

vzakhari replied:
Thanks! I will use the greedy rewriter.

jeanPerier reviewed (approved):
Looks great, those helper operations makes sense to me.

tblah commented on flang/lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -4671,6 +4671,79 @@ void fir::UnpackArrayOp::getEffects(
+};
+} // namespace
+
+mlir::LogicalResult SimplifyBoxTotalElementsOp::matchAndRewrite(
```
I think we could also canonicalize this for cases where the number of elements is known at compile time inside of the type e.g. !fir.box<!fir.array<10xi32>>

tblah reviewed (commented):
Thanks for this. Why did you decide to lower these ops in a separate pass instead of in early codegen e.g. codegen-rewrite?

vzakhari commented:
> Thanks for this. Why did you decide to lower these ops in a separate pass instead of in early codegen e.g. codegen-rewrite?

I plan to use FIR loops (and maybe SCF ifs) for expanding these ops, so this has to be done before SCF-to-CF conversion, thus, before codegen-rewrite.  I guess I could try to use just `arith.select`'s, but that the loops will be more compact and clear.

vzakhari committed:
Changed `IsContiguousUpTo` to accept int dim argument.

vzakhari committed:
Clarified that the input box cannot be absent.
I do not want to identify later whether the box might be absent
and generate redundant present checks. To avoid redundant present
checks I can add attribute like `present`/`maybe_absent`,
but I think it should be easy to generate these ops under
the present check upstream. So I would like to make them
simple at least for now.

vzakhari committed:
Use greedy rewriter.

vzakhari committed:
Added TODOs for more canonicalization.

vzakhari commented:
I hope you do not mind that I just put TODOs in the canonicalization area.  I would like to focus on array repacking, and then get back to handling more canonicalization, and using new operations earlier in the compilation (e.g. in intrinsic lowering).

jeanPerier reviewed (approved):
Thanks for the update, looks great!
